text/0000-optimise-attr.md

```diff
@@ -0,0 +1,120 @@
+- Feature Name: optimise_attr
+- Start Date: 2018-03-26
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+This RFC introduces the `#[optimise]` attribute, specifically its `#[optimise(size)]` variant for
+controlling optimisation level on a per-item basis.
+
+# Motivation
+[motivation]: #motivation
+
+Currently, rustc has only a small number of optimisation options that apply globally to the
+crate. With LTO and RLIB-only crates these options become applicable to a whole-program, which
+reduces the ability to control optimisation even further.
+
+For applications such as embedded, it is critical, that they satisfy the size constraints. This
+means, that code must consciously pick one or the other optimisation level. However, since
+optimisation level is increasingly applied program-wide, options like `-Copt-level=3` or
+`-Copt-level=s` are less and less useful – it is no longer feasible (and never was feasible with
+cargo) to use the former one for code where performance matters and the latter everywhere else.
+
+With a C toolchain this is fairly easy to achieve by compiling the relevant objects with different
+options. In Rust ecosystem, however, where this concept does not exist, an alternate solution is
+necessary.
+
+With `#[optimise(size)]` it is possible to annotate separate functions, so that they are optimised
+for size in a project otherwise optimised for speed (which is the default for `cargo --release`).
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Sometimes, optimisations are a tradeoff between execution time and the code size. Some
+optimisations, such as loop unrolling increase code size many times on average (compared to
+original function size).
+
+```rust
+#[optimise(size)]
+fn banana() {
```

Comment:
This function is monomorphic. How do you deal with polymorphic functions? Does

Reply:
Where attributes are involved, monomorphisations behave as if they were monomorphic functions annotated with the same attributes as the generic function they were derived from.

This is generally true, so I didn’t see too much reason to explicitly note this in RFC.